processing/src/main/java/io/druid/query/extraction/FunctionalExtraction.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ * Functional extraction uses a function to find the new value.
+ * null values in the range can either retain the domain's value, or replace the null value with "replaceMissingValueWith"
+ */
+public abstract class FunctionalExtraction extends DimExtractionFn
+{
+  private final boolean retainMissingValue;
+  private final String replaceMissingValueWith;
+  private final Function<String, String> extractionFunction;
+  private final ExtractionType extractionType;
+
+  /**
+   * The general constructor which handles most of the logic for extractions which can be expressed as a function of string-->string
+   *
+   * @param extractionFunction      The function to call when doing the extraction. The function must be able to accept a null input.
+   * @param retainMissingValue      Boolean value on if functions which result in `null` should use the original value or should be kept as `null`
+   * @param replaceMissingValueWith String value to replace missing values with (instead of `null`)
+   * @param uniqueProjections       If this function always has 1:1 renames and the domain has the same cardinality of the input, this should be true and enables optimized query paths.
+   */
+  public FunctionalExtraction(
+      final Function<String, String> extractionFunction,
+      final Boolean retainMissingValue,
+      final String replaceMissingValueWith,
+      final Boolean uniqueProjections
```

Comment:
since we call this "one to one" everywhere else, should we adopt the same term here as well?
Otherwise, if we think in terms of mathematical terms, we could call it `injective`?